### ISSUE TITLE
Empty output when using in Spring's AbstractJackson2HttpMessageConverter

### ISSUE THREAD

**Issue by kurbatov** [michel-kraemer/bson4jackson#62]
I try using bson4jackson in Spring MVC web application to convert domain objects at a REST endpoint.

I define bson convertor as follows:

```java
public class Jackson2BsonHttpMessageConverter extends AbstractJackson2HttpMessageConverter {
    public Jackson2BsonHttpMessageConverter() {
        super(new ObjectMapper(new BsonFactory()), new MediaType("application", "bson"));
    }
}
```

Spring's `AbstractJackson2HttpMessageConverter` then converts a domain object in the following way:

```java
protected void writeInternal(Object object, Type type, HttpOutputMessage outputMessage) {
    JsonGenerator generator = this.objectMapper.getFactory().createGenerator(outputMessage.getBody(), encoding);
...
    ObjectWriter objectWriter = this.objectMapper.writer();
...
    objectWriter.writeValue(generator, value);
...
}
```

And that results in empty output.

I recreated this behavior in a unit-test:

```java
@Test
public class Bson4JacksonTest {

    private static final Logger LOGGER = LoggerFactory.getLogger(Bson4JacksonTest.class);

    @Test
    public void bsonSerializationTest() throws IOException {
        TestPerson testObject = new TestPerson(12L, "Joe", "+79201112233", "84952334455");
        ObjectMapper om = new ObjectMapper(new BsonFactory());
        byte[] result1 = om.writeValueAsBytes(testObject); // returns 89 bytes, looks ok
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        JsonGenerator g = om.getFactory().createGenerator(os);
        om.writeValue(g, testObject);
        g.flush();
        byte[] result2 = os.toByteArray(); // returns 0 bytes, looks wrong
        LOGGER.debug("Bson representation: {}", new String(result2));
        assertTrue(result2.length > 0); // fails
        assertEquals(result2, result1); // fails
    }

    public static class TestPerson {
        private Long id;
        private String name;
        private Set<String> phoneNumbers;
        // constructors and accessors are ommited
    }
}
```

Unfortunately, I cannot debug it deep enough to find a root cause.

I think that it is a bug because the same approach of using Spring's `AbstractJackson2HttpMessageConverter` works with [jackson-dataformat-msgpack](https://github.com/msgpack/msgpack-java/tree/develop/msgpack-jackson) library that integrates to jackson in the same way.

Could you please fix that behavior or suggest a workaround?

I have tried bson4jackson 2.6.0 and 2.7.0 with jackson 2.6.3, 2.6.4, 2.6.5, 2.7.0 and 2.7.3

**Comment by lyind** [comment]
I think I encountered the same issue and fixed it:

https://github.com/michel-kraemer/bson4jackson/pull/63/commits/b599f5c409c3f1c01ebca06bcb2ac7a6d1613f0b

**Comment by michel-kraemer** [comment]
@kurbatov it would be awesome if you could build bson4jackson with the fix from @lyind and test if it really fixes your problem. I can then merge the PR and release a new version.

**Comment by michel-kraemer** [comment]
@kurbatov  I hope that's not too much to ask. If you don't have the time to test the fix, I can completely understand it. I can merge the PR anyway, but I still have to test it myself, and your help would be much appreciated.

**Comment by kurbatov** [comment]
@michel-kraemer it worked. I tested using with a unit-test I quoted above. Maybe it make sence to add it to project's unit tests.

**Comment by kurbatov** [comment]
It worked in my original spring setup as well.
@lyind , @michel-kraemer thank you.

**Comment by michel-kraemer** [comment]
Ah perfect. Thank you! @lyind may I ask you to add the unit test to your PR?

**Comment by lyind** [comment]
I will add a test.

**Comment by michel-kraemer** [comment]
I just merged #63. The issue should be fixed.
